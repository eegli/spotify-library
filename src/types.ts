import { AxiosResponse } from 'axios';

type PickActual<T, P extends keyof T> = T[P];

export type AppConfig = {
  token: string;
  type: 'full' | 'normal';
  genres: boolean;
  features: boolean;
  outDir: string;
};

export type TrackLight = {
  id: string;
  name: string;
  album: {
    name: string;
    id: string;
  };
  artists: {
    id: string;
    name: string;
  }[];
};

export type TrackFull = PickActual<SpotifyApi.SavedTrackObject, 'track'>;

export type Library<T> = {
  added_at: string;
  track: T & {
    genres?: string[][];
    features?: SpotifyApi.AudioFeaturesObject;
  };
}[];

export type UsersSavedTracks =
  AxiosResponse<SpotifyApi.UsersSavedTracksResponse>;
export type MultipleArtists = AxiosResponse<SpotifyApi.MultipleArtistsResponse>;
export type MultipleAudioFeatures =
  AxiosResponse<SpotifyApi.MultipleAudioFeaturesResponse>;
