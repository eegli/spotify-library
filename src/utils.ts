import cliProgress from 'cli-progress';
import fs from 'fs';
import { join } from 'path';

export function hasOwnProperty<
  T extends Record<PropertyKey, unknown>,
  K extends PropertyKey
>(obj: T, prop: K): obj is T & Record<K, unknown> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

export function write(path: string, fileName: string, data: unknown): string {
  path = join(process.cwd(), path);
  if (!fs.existsSync(path)) {
    fs.mkdirSync(path, { recursive: true });
  }
  if (fileName.endsWith('.json')) {
    fileName = fileName.slice(0, -5);
  }
  path = join(path, fileName + '.json');
  fs.writeFileSync(path, JSON.stringify(data, null, 2));
  return path;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function chunkify<T extends any[] | Set<any>>(
  obj: T,
  chunkSize: number
): T[][] {
  const len = Array.isArray(obj) ? obj.length : obj.size;
  const chunks = [];
  for (let i = 0; i < len; i += chunkSize) {
    chunks.push([...obj].slice(i, i + chunkSize));
  }
  return chunks;
}

export function createProgressBar(items: string) {
  return new cliProgress.SingleBar({
    format:
      `Fetching ${items}\t |` +
      '{bar} {percentage}% | ETA: {eta}s | {value}/{total}',
    barCompleteChar: '\u25A0',
    barIncompleteChar: ' ',
    hideCursor: true,
  });
}

export function goodbye(message: string): never {
  console.error('\n', '\x1b[31m', `Error: ${message}. Goodbye`, '\x1b[0m');
  process.exit(1);
}
