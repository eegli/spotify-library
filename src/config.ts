import { AppConfig } from './types';
import { goodbye, hasOwnProperty } from './utils';

export const defaultConfig: AppConfig = {
  token: '',
  type: 'normal',
  genres: false,
  features: false,
  outDir: '',
};

export const createConfig = (args: string[]): AppConfig => {
  const config: AppConfig = { ...defaultConfig };

  args.forEach((val, idx, orig) => {
    if (val.startsWith('--')) {
      const key = val.slice(2);
      if (hasOwnProperty(config, key)) {
        // Flag argument
        if (orig[idx + 1] && !orig[idx + 1].startsWith('--')) {
          config[key] = orig[idx + 1];
          // Boolean flag
        } else {
          config[key] = true;
        }
      } else {
        console.warn(`Ignoring unknown option: ${val}`);
      }
    }
  });

  if (!['full', 'normal'].includes(config.type)) {
    config.type = defaultConfig.type;
  }

  if (!config.token) {
    goodbye('Missing Spotify authentication token');
  }

  return config;
};
