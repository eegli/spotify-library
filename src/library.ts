import axios from 'axios';
import { parseConfig } from './config';
import type {
  Library,
  MultipleArtists,
  MultipleAudioFeatures,
  TrackFull,
  TrackLight,
  UsersSavedTracks,
} from './types';
import { chunkify, createProgressBar, goodbye, write } from './utils';

export async function library(): Promise<Library<TrackLight | TrackFull>> {
  const config = await parseConfig(process.argv.slice(2));
  try {
    let library: Library<TrackFull | TrackLight> = [];
    let progress = createProgressBar('user library');

    const pagination = {
      nextUrl: <string | null>'https://api.spotify.com/v1/me/tracks',
      start: 0,
      end: 50,
    };

    progress.start(0, 0);
    while (pagination.nextUrl) {
      const { data }: UsersSavedTracks = await axios.get(pagination.nextUrl, {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
        params: {
          limit: 50,
        },
      });

      progress.setTotal(data.total);
      progress.increment(data.items.length);

      data.items.forEach(el => {
        library.push(el);
      });

      pagination.nextUrl = data.next;
      pagination.start = pagination.end;
      pagination.end += data.limit;
    }

    progress.stop();

    // Reduce library if necessary
    if (config.type === 'normal') {
      library = library.reduce((acc, curr) => {
        acc.push({
          added_at: curr.added_at,
          track: {
            id: curr.track.id,
            name: curr.track.name,
            album: {
              id: curr.track.album.id,
              name: curr.track.album.name,
            },
            artists: curr.track.artists.map(a => ({
              id: a.id,
              name: a.name,
            })),
          },
        });
        return acc;
      }, <Library<TrackLight>>[]);
    }

    // Add genres if specified
    if (config.genres) {
      const artists = library.map(t => t.track.artists.map(a => a.id)).flat();
      const artistIds = new Set<string>(artists);
      const genres: Record<string, string[]> = {};

      // Max 50 items per request to Spotify
      const chunkSize = 50;

      progress = createProgressBar('artists');
      progress.start(artistIds.size, 0);

      for await (const chunk of chunkify(artistIds, chunkSize)) {
        const { data }: MultipleArtists = await axios.get(
          'https://api.spotify.com/v1/artists',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (config.features) {
      const songIds = library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      const chunkSize = 50;
      progress = createProgressBar('audio features');

      progress.start(songIds.length, 0);
      for await (const chunk of chunkify(songIds, 50)) {
        const { data }: MultipleAudioFeatures = await axios.get(
          'https://api.spotify.com/v1/audio-features',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );

        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const outDir = write(config.outDir, 'spotify-library', library);
    console.info("Success! Library written to '%s'", outDir);
    return library;
  } catch (e) {
    goodbye('Something went wrong: ' + e);
  }
}
