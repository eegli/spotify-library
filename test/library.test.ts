import axios from 'axios';
import { library } from '../src/library';
import * as utils from '../src/utils';
import {
  RES_MULTIPLE_ARTISTS,
  RES_MULTIPLE_AUDIO_FEATURES,
  RES_USER_SAVED_TRACKS,
} from './fixtures';

const writeSpy = jest.spyOn(utils, 'write');
const mockAxios = axios as jest.Mocked<typeof axios>;

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Library', () => {
  it('basic library', async () => {
    process.argv = ['', '', '--token', '123', '--type', 'normal'];
    mockAxios.get.mockResolvedValueOnce(RES_USER_SAVED_TRACKS);
    const res = await library();
    expect(mockAxios.get).toHaveBeenCalledTimes(1);
    expect(res).toMatchSnapshot();
    expect(writeSpy).toHaveBeenCalledTimes(1);
  });
  it('with genres', async () => {
    process.argv = ['', '', '--token', '123', '--genres'];
    mockAxios.get
      .mockResolvedValueOnce(RES_USER_SAVED_TRACKS)
      .mockResolvedValueOnce(RES_MULTIPLE_ARTISTS);
    const res = await library();
    expect(mockAxios.get).toHaveBeenCalledTimes(2);
    expect(res).toMatchSnapshot();
  });
  it('with audio features', async () => {
    process.argv = ['', '', '--token', '123', '--features'];
    mockAxios.get
      .mockResolvedValueOnce(RES_USER_SAVED_TRACKS)
      .mockResolvedValueOnce(RES_MULTIPLE_AUDIO_FEATURES);
    const res = await library();
    expect(mockAxios.get).toHaveBeenCalledTimes(2);
    expect(res).toMatchSnapshot();
  });
  it('with all', async () => {
    process.argv = ['', '', '--token', '123', '--features', '--genres'];
    mockAxios.get
      .mockResolvedValueOnce(RES_USER_SAVED_TRACKS)
      .mockResolvedValueOnce(RES_MULTIPLE_ARTISTS)
      .mockResolvedValueOnce(RES_MULTIPLE_AUDIO_FEATURES);
    const res = await library();
    expect(mockAxios.get).toHaveBeenCalledTimes(3);
    expect(res).toMatchSnapshot();
  });
  it('throws', async () => {
    process.argv = ['', '', '--token', '123'];
    // No axios response
    await expect(library()).rejects.toThrow();
  });
});
