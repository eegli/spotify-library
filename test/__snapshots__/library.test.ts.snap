// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library basic library 1`] = `
Array [
  Object {
    "added_at": "2021-09-30T22:04:57Z",
    "track": Object {
      "album": Object {
        "id": "0vlWyzbAl3ff9b3zout1XQ",
        "name": "Stay With Me",
      },
      "artists": Array [
        Object {
          "id": "5Ho1vKl1Uz8bJlk4vbmvmf",
          "name": "Amelie Lens",
        },
      ],
      "id": "5UsfWcP6SThHlZ4oAgx7ge",
      "name": "Follow",
    },
  },
  Object {
    "added_at": "2021-12-06T17:17:57Z",
    "track": Object {
      "album": Object {
        "id": "6Ozmnoay8pclUfUO8CWKLU",
        "name": "The Age Of Love (Charlotte de Witte & Enrico Sangiuliano Remix)",
      },
      "artists": Array [
        Object {
          "id": "03tes5RhEvH2dX2eDeGnRn",
          "name": "Age Of Love",
        },
        Object {
          "id": "1lJhME1ZpzsEa5M0wW6Mso",
          "name": "Charlotte de Witte",
        },
        Object {
          "id": "1u7DsNFbakULvxnDGtMm90",
          "name": "Enrico Sangiuliano",
        },
      ],
      "id": "6X6Ow0iioCrYQSK1QDHQbC",
      "name": "The Age Of Love - Charlotte de Witte & Enrico Sangiuliano Remix",
    },
  },
  Object {
    "added_at": "2021-10-18T08:15:08Z",
    "track": Object {
      "album": Object {
        "id": "046wH7DEfUumC5h9ykIEkd",
        "name": "bb u ok?",
      },
      "artists": Array [
        Object {
          "id": "0jNDKefhfSbLR9sFvcPLHo",
          "name": "San Holo",
        },
        Object {
          "id": "1BjaGDkxwa2fb2pSCXlFXb",
          "name": "Chet Porter",
        },
      ],
      "id": "3LnoLo62GjfKDYIN7eJpPB",
      "name": "you’ve changed, i’ve changed",
    },
  },
]
`;

exports[`Library with all 1`] = `
Array [
  Object {
    "added_at": "2021-09-30T22:04:57Z",
    "track": Object {
      "album": Object {
        "id": "0vlWyzbAl3ff9b3zout1XQ",
        "name": "Stay With Me",
      },
      "artists": Array [
        Object {
          "id": "5Ho1vKl1Uz8bJlk4vbmvmf",
          "name": "Amelie Lens",
        },
      ],
      "features": Object {
        "acousticness": 0.0375,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/5UsfWcP6SThHlZ4oAgx7ge",
        "danceability": 0.725,
        "duration_ms": 337500,
        "energy": 0.908,
        "id": "5UsfWcP6SThHlZ4oAgx7ge",
        "instrumentalness": 0.822,
        "key": 7,
        "liveness": 0.108,
        "loudness": -8.798,
        "mode": 1,
        "speechiness": 0.0345,
        "tempo": 128.006,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/5UsfWcP6SThHlZ4oAgx7ge",
        "type": "audio_features",
        "uri": "spotify:track:5UsfWcP6SThHlZ4oAgx7ge",
        "valence": 0.134,
      },
      "genres": Array [
        Array [
          "acid techno",
          "belgian techno",
          "dark techno",
          "electra",
        ],
      ],
      "id": "5UsfWcP6SThHlZ4oAgx7ge",
      "name": "Follow",
    },
  },
  Object {
    "added_at": "2021-12-06T17:17:57Z",
    "track": Object {
      "album": Object {
        "id": "6Ozmnoay8pclUfUO8CWKLU",
        "name": "The Age Of Love (Charlotte de Witte & Enrico Sangiuliano Remix)",
      },
      "artists": Array [
        Object {
          "id": "03tes5RhEvH2dX2eDeGnRn",
          "name": "Age Of Love",
        },
        Object {
          "id": "1lJhME1ZpzsEa5M0wW6Mso",
          "name": "Charlotte de Witte",
        },
        Object {
          "id": "1u7DsNFbakULvxnDGtMm90",
          "name": "Enrico Sangiuliano",
        },
      ],
      "features": Object {
        "acousticness": 0.0576,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/6X6Ow0iioCrYQSK1QDHQbC",
        "danceability": 0.748,
        "duration_ms": 489480,
        "energy": 0.909,
        "id": "6X6Ow0iioCrYQSK1QDHQbC",
        "instrumentalness": 0.89,
        "key": 5,
        "liveness": 0.0945,
        "loudness": -8.322,
        "mode": 0,
        "speechiness": 0.0505,
        "tempo": 130,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/6X6Ow0iioCrYQSK1QDHQbC",
        "type": "audio_features",
        "uri": "spotify:track:6X6Ow0iioCrYQSK1QDHQbC",
        "valence": 0.0617,
      },
      "genres": Array [
        Array [
          "bubble trance",
        ],
        Array [
          "belgian techno",
          "electra",
        ],
        Array [
          "dark techno",
          "italian techno",
          "minimal techno",
        ],
      ],
      "id": "6X6Ow0iioCrYQSK1QDHQbC",
      "name": "The Age Of Love - Charlotte de Witte & Enrico Sangiuliano Remix",
    },
  },
  Object {
    "added_at": "2021-10-18T08:15:08Z",
    "track": Object {
      "album": Object {
        "id": "046wH7DEfUumC5h9ykIEkd",
        "name": "bb u ok?",
      },
      "artists": Array [
        Object {
          "id": "0jNDKefhfSbLR9sFvcPLHo",
          "name": "San Holo",
        },
        Object {
          "id": "1BjaGDkxwa2fb2pSCXlFXb",
          "name": "Chet Porter",
        },
      ],
      "features": Object {
        "acousticness": 0.123,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/3LnoLo62GjfKDYIN7eJpPB",
        "danceability": 0.416,
        "duration_ms": 311625,
        "energy": 0.615,
        "id": "3LnoLo62GjfKDYIN7eJpPB",
        "instrumentalness": 0.00171,
        "key": 5,
        "liveness": 0.273,
        "loudness": -7.898,
        "mode": 1,
        "speechiness": 0.0452,
        "tempo": 130.228,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/3LnoLo62GjfKDYIN7eJpPB",
        "type": "audio_features",
        "uri": "spotify:track:3LnoLo62GjfKDYIN7eJpPB",
        "valence": 0.181,
      },
      "genres": Array [
        Array [
          "edm",
          "electronic trap",
          "electropop",
          "future bass",
          "pop dance",
          "pop edm",
          "vapor twitch",
        ],
        Array [
          "vapor twitch",
        ],
      ],
      "id": "3LnoLo62GjfKDYIN7eJpPB",
      "name": "you’ve changed, i’ve changed",
    },
  },
]
`;

exports[`Library with audio features 1`] = `
Array [
  Object {
    "added_at": "2021-09-30T22:04:57Z",
    "track": Object {
      "album": Object {
        "id": "0vlWyzbAl3ff9b3zout1XQ",
        "name": "Stay With Me",
      },
      "artists": Array [
        Object {
          "id": "5Ho1vKl1Uz8bJlk4vbmvmf",
          "name": "Amelie Lens",
        },
      ],
      "features": Object {
        "acousticness": 0.0375,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/5UsfWcP6SThHlZ4oAgx7ge",
        "danceability": 0.725,
        "duration_ms": 337500,
        "energy": 0.908,
        "id": "5UsfWcP6SThHlZ4oAgx7ge",
        "instrumentalness": 0.822,
        "key": 7,
        "liveness": 0.108,
        "loudness": -8.798,
        "mode": 1,
        "speechiness": 0.0345,
        "tempo": 128.006,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/5UsfWcP6SThHlZ4oAgx7ge",
        "type": "audio_features",
        "uri": "spotify:track:5UsfWcP6SThHlZ4oAgx7ge",
        "valence": 0.134,
      },
      "id": "5UsfWcP6SThHlZ4oAgx7ge",
      "name": "Follow",
    },
  },
  Object {
    "added_at": "2021-12-06T17:17:57Z",
    "track": Object {
      "album": Object {
        "id": "6Ozmnoay8pclUfUO8CWKLU",
        "name": "The Age Of Love (Charlotte de Witte & Enrico Sangiuliano Remix)",
      },
      "artists": Array [
        Object {
          "id": "03tes5RhEvH2dX2eDeGnRn",
          "name": "Age Of Love",
        },
        Object {
          "id": "1lJhME1ZpzsEa5M0wW6Mso",
          "name": "Charlotte de Witte",
        },
        Object {
          "id": "1u7DsNFbakULvxnDGtMm90",
          "name": "Enrico Sangiuliano",
        },
      ],
      "features": Object {
        "acousticness": 0.0576,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/6X6Ow0iioCrYQSK1QDHQbC",
        "danceability": 0.748,
        "duration_ms": 489480,
        "energy": 0.909,
        "id": "6X6Ow0iioCrYQSK1QDHQbC",
        "instrumentalness": 0.89,
        "key": 5,
        "liveness": 0.0945,
        "loudness": -8.322,
        "mode": 0,
        "speechiness": 0.0505,
        "tempo": 130,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/6X6Ow0iioCrYQSK1QDHQbC",
        "type": "audio_features",
        "uri": "spotify:track:6X6Ow0iioCrYQSK1QDHQbC",
        "valence": 0.0617,
      },
      "id": "6X6Ow0iioCrYQSK1QDHQbC",
      "name": "The Age Of Love - Charlotte de Witte & Enrico Sangiuliano Remix",
    },
  },
  Object {
    "added_at": "2021-10-18T08:15:08Z",
    "track": Object {
      "album": Object {
        "id": "046wH7DEfUumC5h9ykIEkd",
        "name": "bb u ok?",
      },
      "artists": Array [
        Object {
          "id": "0jNDKefhfSbLR9sFvcPLHo",
          "name": "San Holo",
        },
        Object {
          "id": "1BjaGDkxwa2fb2pSCXlFXb",
          "name": "Chet Porter",
        },
      ],
      "features": Object {
        "acousticness": 0.123,
        "analysis_url": "https://api.spotify.com/v1/audio-analysis/3LnoLo62GjfKDYIN7eJpPB",
        "danceability": 0.416,
        "duration_ms": 311625,
        "energy": 0.615,
        "id": "3LnoLo62GjfKDYIN7eJpPB",
        "instrumentalness": 0.00171,
        "key": 5,
        "liveness": 0.273,
        "loudness": -7.898,
        "mode": 1,
        "speechiness": 0.0452,
        "tempo": 130.228,
        "time_signature": 4,
        "track_href": "https://api.spotify.com/v1/tracks/3LnoLo62GjfKDYIN7eJpPB",
        "type": "audio_features",
        "uri": "spotify:track:3LnoLo62GjfKDYIN7eJpPB",
        "valence": 0.181,
      },
      "id": "3LnoLo62GjfKDYIN7eJpPB",
      "name": "you’ve changed, i’ve changed",
    },
  },
]
`;

exports[`Library with genres 1`] = `
Array [
  Object {
    "added_at": "2021-09-30T22:04:57Z",
    "track": Object {
      "album": Object {
        "id": "0vlWyzbAl3ff9b3zout1XQ",
        "name": "Stay With Me",
      },
      "artists": Array [
        Object {
          "id": "5Ho1vKl1Uz8bJlk4vbmvmf",
          "name": "Amelie Lens",
        },
      ],
      "genres": Array [
        Array [
          "acid techno",
          "belgian techno",
          "dark techno",
          "electra",
        ],
      ],
      "id": "5UsfWcP6SThHlZ4oAgx7ge",
      "name": "Follow",
    },
  },
  Object {
    "added_at": "2021-12-06T17:17:57Z",
    "track": Object {
      "album": Object {
        "id": "6Ozmnoay8pclUfUO8CWKLU",
        "name": "The Age Of Love (Charlotte de Witte & Enrico Sangiuliano Remix)",
      },
      "artists": Array [
        Object {
          "id": "03tes5RhEvH2dX2eDeGnRn",
          "name": "Age Of Love",
        },
        Object {
          "id": "1lJhME1ZpzsEa5M0wW6Mso",
          "name": "Charlotte de Witte",
        },
        Object {
          "id": "1u7DsNFbakULvxnDGtMm90",
          "name": "Enrico Sangiuliano",
        },
      ],
      "genres": Array [
        Array [
          "bubble trance",
        ],
        Array [
          "belgian techno",
          "electra",
        ],
        Array [
          "dark techno",
          "italian techno",
          "minimal techno",
        ],
      ],
      "id": "6X6Ow0iioCrYQSK1QDHQbC",
      "name": "The Age Of Love - Charlotte de Witte & Enrico Sangiuliano Remix",
    },
  },
  Object {
    "added_at": "2021-10-18T08:15:08Z",
    "track": Object {
      "album": Object {
        "id": "046wH7DEfUumC5h9ykIEkd",
        "name": "bb u ok?",
      },
      "artists": Array [
        Object {
          "id": "0jNDKefhfSbLR9sFvcPLHo",
          "name": "San Holo",
        },
        Object {
          "id": "1BjaGDkxwa2fb2pSCXlFXb",
          "name": "Chet Porter",
        },
      ],
      "genres": Array [
        Array [
          "edm",
          "electronic trap",
          "electropop",
          "future bass",
          "pop dance",
          "pop edm",
          "vapor twitch",
        ],
        Array [
          "vapor twitch",
        ],
      ],
      "id": "3LnoLo62GjfKDYIN7eJpPB",
      "name": "you’ve changed, i’ve changed",
    },
  },
]
`;
